### PR TITLE
Add Heraldicon integration

### PR DESCRIPTION
**Pull request by or**
For https://github.com/mirnovov/heraldtron/issues/1

Apologies for the noise, I noticed there was a lot of trailing whitespace, I thought it prudent to get that out of the way to avoid conflicts or noise in later commits.

Also noticed a trailing comma in the schema.

The final commit should be easier to review on its own, tho.

### CODE REVIEW

File: ht/services.py [hunk 1]
```diff
@@ -38,39 +38,126 @@ def image_result(item):
...
+async def heraldicon(session, query):
+	options, blazon = parse_options_and_blazon(query)
+	result = await utils.post_json(session, "https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api",
+					{
+						"call": "generate-from-blazon",
+						"data": {
+							"blazon": blazon,
+							"options": options,
+						},
+					})
```
Comment (mirnovov) [inline ht/services.py]:
```suggestion
	result = await utils.post_json(session, "https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api",
		{
			"call": "generate-from-blazon",
			"data": {
				"blazon": blazon,
				"options": options,
			},
		}
	)
```
More in line with my standard code style

File: ht/services.py [hunk 2]
```diff
@@ -38,39 +38,126 @@ def image_result(item):
...
+	success = result.get("success")
+	if success:
+		png_url = success["png-url"]
+		edit_url = success['edit-url']
+		embed = embeds.DRAW.create("", edit_url, heading = "Success")
+		embed.set_thumbnail(url = "https://cdn.heraldicon.org/img/heraldicon-logo.png")
+		embed.set_image(url = png_url)
+		embed.add_field(name = "blazon", value = f"*{blazon}*", inline = True)
+		embed.set_footer(text = "CC BY-SA 4.0, attribution on https://heraldicon.org")
```
Comment (mirnovov) [inline ht/services.py]:
```suggestion
		png_url = success["png-url"]
		edit_link = f"[View in Heraldicon]({success['edit-url']})"
		embed = embeds.DRAW.create("", edit_link, heading = "Shield created!")
		embed.set_image(url = png_url)
		embed.add_field(name = "Blazon", value = f"*{blazon}*", inline = True)
		embed.set_footer(
			icon_url = "https://cdn.heraldicon.org/img/heraldicon-logo.png",
			text = "Drawn using Heraldicon; licensed under CC BY-SA 4.0 (attribution on https://heraldicon.org)"
		)
```
A few changes here:
- "Success" is kinda programmer-y lol, this is more consistent with the DS command
- I understand that you want to advertise your website, but the icon on the side is simply too big imo and kinda distracting if someone's doing a whole bunch of commands. I've moved it to the footer.
- Replaced the raw url with a link, looks a bit nicer imo

File: ht/services.py [hunk 3]
```diff
@@ -38,39 +38,126 @@ def image_result(item):
...
+	embed.set_thumbnail(url = "https://cdn.heraldicon.org/img/heraldicon-logo.png")
+
+	return embed
+
+async def heraldicon_options(session):
+	result = await utils.post_json(session, "https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api",
+					{
+						"call": "blazon-options",
+					})
+
```
Comment (mirnovov) [inline ht/services.py]:
```suggestion
	result = await utils.post_json(
		session, 
		"https://2f1yb829vl.execute-api.eu-central-1.amazonaws.com/api",
		{"call": "blazon-options"}
	)
```
again, nicer formatting imo

File: ht/services.py [hunk 4]
```diff
@@ -38,39 +38,126 @@ def image_result(item):
...
+		embed.add_field(name = "theme", value = ", ".join(f"`:{x}`" for x in themes), inline = False)
+		textures = result["success"]["options"]["texture"]
+		embed.add_field(name = "texture", value = ", ".join(f"`:{x}`" for x in textures), inline = False)
+
+		return embed
+
+	embed = embeds.DRAW.create("", "Unknown error.", heading = "Error")
+	embed.set_thumbnail(url = "https://cdn.heraldicon.org/img/heraldicon-logo.png")
+
+	return embed
```
Comment (mirnovov) [inline ht/services.py]:
```suggestion
		embed = embeds.DRAW.create(
			"", 
			"Put any of these options as first words before the blazon.", 
			heading = "Heraldicon rendering options"
		)
				
		add_option_type(embed, result["success"]["options"]["miscellaneous"], "General")
		add_option_type(embed, result["success"]["options"]["mode"], "Mode")
		add_option_type(embed, result["success"]["options"]["escutcheon"], "Escutcheons")
		add_option_type(embed, result["success"]["options"]["theme"], "Themes")
		add_option_type(embed, result["success"]["options"]["texture"], "Textures")
		
		embed.set_footer(
			icon_url = "https://cdn.heraldicon.org/img/heraldicon-logo.png",
			text = "For more info, see https://heraldicon.org"
		)
		return embed

	raise utils.CustomCommandError(
		"Cannot retrieve Heraldicon options",
		"An unknown error occured while trying to retrieve the Heraldicon options."
	)
	
def add_option_type(embed, source, name):
	embed.add_field(
		name = name, 
		value = ", ".join(f"`:{x}`" for x in source), 
		inline = False
	)
```

- Killed repeat code by moving the embed fields into a function (this is done elsewhere in the bot as well)
- The bot uses a standard error mechanism throughout the code, so replaced your error embed with that
- Again, relegate the icon to the footer

**Comment by mirnovov** [comment]
Anyway, thanks very much for actually getting your hands dirty and adding the code haha, hopefully my changes aren't too much of an ask

**Comment by or** [comment]
Groovy!
- I noticed a little bug with the error display, fixed that.
- Changed link text to "Edit on Heraldicon".
- Moved the `add_option_type` above `heraldicon_options`, so it is defined before.
- Fixed a spelling mistake.

The icon in the footer is a good idea, I didn't see that that can be done. I mainly wanted it to distinguish the responses from the DS ones, but it's fine down there as well.

**Comment by mirnovov** [comment]
Thanks very much!

**Pull request merged**